### PR TITLE
Remove redundant static typing casts

### PR DESCRIPTION
**Pull request by danieleades**
remove redundant `typing.cast`s and updates mypy config to prevent these in future.

type casts should only be used as a last resort, as they can mask errors if refactoring later changes the type of an object.

### CODE REVIEW

**Comment by danieleades** [comment]
i don't believe the CI failure is to do with this PR

File: setup.cfg [hunk 1]
```diff
@@ -32,7 +32,7 @@ follow_imports = skip
 check_untyped_defs = True
 warn_unused_ignores = True
 strict_optional = False
-no_implicit_optional = True
```
Comment (tk0miya) [inline setup.cfg:35]: Why did you remove this line?

File: sphinx/domains/c.py [hunk 2]
```diff
@@ -3552,7 +3552,6 @@ def _render_symbol(self, s: Symbol, maxdepth: int, skipThis: bool,
 
     def apply(self, **kwargs: Any) -> None:
         for node in self.document.findall(AliasNode):
-            node = cast(AliasNode, node)
```
Comment (tk0miya) [inline sphinx/domains/c.py:3555]: Good point. Our stubs surely provide nice annotation for `findall()`!
https://github.com/tk0miya/docutils-stubs/blob/b14071d065f8ee35399856fcaff6cd608294a342/docutils-stubs/nodes.pyi#L32-L33

**Review by tk0miya — COMMENTED** [review]
LGTM with nits.

**Comment by tk0miya** [comment]
I manually re-run failed tests :-)

File: setup.cfg [hunk 3]
```diff
@@ -32,7 +32,7 @@ follow_imports = skip
...
+warn_redundant_casts = True
```
Comment (AA-Turner) [inline setup.cfg:36]: Please discuss the config change in a different PR/issue.
```suggestion
no_implicit_optional = True
warn_redundant_casts = True
```

**Review by AA-Turner — APPROVED** [review]

**Comment by AA-Turner** [comment]
Thanks Dan!

A
